backend/molgenis-emx2-webapi/src/main/java/org/molgenis/emx2/web/MolgenisSessionManager.java

```diff
@@ -7,97 +7,78 @@
+      if (sessions.containsKey(token)) {
+        return sessions.get(token);
       }
     }
 
-    // otherwise try tokens (also in case of sessionless requests)
-    final String user =
-        request.headers(MOLGENIS_TOKEN) == null
-            ? "anonymous"
-            : request.headers(MOLGENIS_TOKEN).replaceAll("[\n|\r|\t]", "_");
-
-    return sessions.computeIfAbsent(
-        user,
-        t -> {
-          Database database = new SqlDatabase(false);
-          if (!database.hasUser(user)) {
-            throw new MolgenisException("Authentication failed: User " + user + " not known");
-          }
-          database.setActiveUser(user);
-          database.setListener(new MolgenisSessionManagerDatabaseListener(this, database));
-          logger.info("Initializing session for user: {}", database.getActiveUser());
-          MolgenisSession session = new MolgenisSession(database);
-          logger.info("Initializing session complete for user: {}", database.getActiveUser());
-          return session;
-        });
-  }
-
-  void updateSession(Request request, Response response) {
-    MolgenisSession session = getSession(request);
-
-    // check if we need to put in session because user has logged in
-    if (session.getDatabase().getActiveUser() != null && request.session() == null) {
+    // if new session create a MolgenisSession object
+    if (request.session().isNew()) {
       request.session(true);
+      // add session into session pool
+      MolgenisSession session = createSession(request.session().id());
+      // put in request session so we can easily access for this session
       request.session().attribute(SESSION_ATTRIBUTE, session);
-      logger.info("Saved session for user: {}", session.getDatabase().getActiveUser());
     }
 
-    // check if session state and session user map still in sync
-    if (session.getSessionUser() != null
-        && !session.getSessionUser().equals(session.getDatabase().getActiveUser())) {
-      // remove old sessions
-      sessions.remove(session.getSessionUser());
-      request.session(false);
-      logger.info("Destroyed session because user {} logged out", session.getSessionUser());
-
-      // only create new session is user != null
-      if (session.getDatabase().getActiveUser() != null) {
-        MolgenisSession newSession = new MolgenisSession(session.getDatabase());
-        request.session(true);
-        request.session().attribute(SESSION_ATTRIBUTE, newSession);
-        sessions.put(newSession.getSessionUser(), newSession);
-        logger.info(
-            "Changed session from old user({}]) to new user({}) because login changed",
-            session.getSessionUser(),
-            newSession.getSessionUser());
-      }
+    // get the session
+    MolgenisSession session = request.session().attribute(SESSION_ATTRIBUTE);
+
+    // check session is invalid
+    if (minutesBetween(session.getCreateTime(), DateTime.now())
```

Comment:
If I understand correctly, the MolgenisSession has its own timeout mechanism, besides the existing servlet engine session? Why is this needed?

Reply:
Good question, the session timeout cannot be changed (unless we go configure Jetty). So this seemed easier. Might be unneeded if jetty timeout is also 30mins. You point to another issue which is to remove timed out sessions from session map.

Reply:
Best leave session timeouts up to the servlet container. I think the servlet API has a session listener interface where you can listen for session timeouts.

Reply:
rewrote to use let jetty take care of the timeouts.